ddR/R/ddR.R

```diff
@@ -178,16 +219,202 @@ setGeneric("do_collect", function(x, parts) standardGeneric("do_collect"))
 #' Homepage: https://github.com/vertica/ddR
 #' @examples
 #' \dontrun{
-#' a <- dlapply(1:5,function(x) x, nparts=3) 
-#' # A DList with 3 partitions, 
+#' a <- dlapply(1:5,function(x) x, nparts=3)
+#' # A DList with 3 partitions,
 #' # which in the aggregate contains the elements 1 through 5.
 #' b <- dlapply(a,function(x) x+3) # AddR 3 to each element of dlist a.
-#' } 
+#' }
 #' @export
 dlapply <- function(X,FUN,...,nparts=NULL) {
    dmapply(FUN,X,MoreArgs=list(...),output.type="dlist",nparts=nparts)
 }
 
+
+#' Validate the arguments for dmapply
+#'
+#' @param ... iterable arguments for dmapply
+#' @return list of validated arguments
+validate_dargs <- function(...){
+    dargs <- list(...)
+    if(length(dargs) == 0) stop("Need to supply at least one iterable item.")
+
+    check_backend_return_length <- function(x){
+        if(is(x,"DObject") && !identical(x@driver, ddR.env$driver))
+            stop(paste0("An argument passed in was created with backend '",
+                        x@driver@name,"'; the currently loaded backend is '",
+                        ddR.env$driver@name,"'."))
+        # length() works correctly for data.frame, arrays, and lists
+        length(x)
+    }
+
+    lens <- vapply(dargs, check_backend_return_length, FUN.VALUE=numeric(1))
+
+    if(max(lens) != min(lens)) stop("The lengths of the iterable arguments need to be equal.")
+    if(min(lens) == 0) stop("Zero-length arguments are not permitted.")
+    dargs
+}
+
+
+# Formerly in do_dmapply method for parallel
+dmapply_helper <- function(driver, func, ..., MoreArgs = list(),
+                   output.type =
+                       c("dlist", "dframe", "darray", "sparse_darray"),
+                   nparts = NULL,
+                   combine = c("default","c", "rbind", "cbind"))
+{
+  stopifnot(is.list(MoreArgs))
+  output.type <- match.arg(output.type)
+  if (!is.null(nparts)) {
+      stopifnot(is.numeric(nparts),
+                length(nparts) == 1L || length(nparts) == 2L)
+  }
+  combine <- match.arg(combine)
+
+  dots <- list(...)
+  dlen<-length(dots)
+
+
+  for(num in 1:dlen){
+    if(is(dots[[num]],"DObject")){
+      #If this is a DObject, we need to extract the backend object and reassemble it (i.e., use collect)
+      #By converting DObject into a normal R object, we let mcmapply handle how iterations on differnt data types
+
+# TODO Clark: The line below needs to change if data is actually
+# to remain distributed
+      dots[[num]] <- collect(dots[[num]])
```

Comment:
I believe this line is the culprit for excessive memory consumption with SNOW, and it should change to support truly distributed data. In general the `do_dmapply` should handle distributed objects. Leaving this for now because it works and it's useful for testing purposes.